### ISSUE TITLE
romberg method issue on scipy >= 1.6.0

### ISSUE THREAD

**Issue by mbrady-2** [scipy/scipy#13691]
My issue is about ...
I use the romberg method in the scipy.integrate module extensively in an application.  The correctness of the output is checked extensively using many automated unit and regression tests.  When upgrading a docker container recently to scipy 1.6.0 I noticed many of my unit tests began failing.  It took 2-3 hours to isolate the issue down to the romberg method breaking in scipy 1.6.0.  The romberg method worked well in scipy 1.5.4 and I've been using the romberg method with no issue for many years.

I have confirmed using an alternative romberg method implementation that the new scipy 1.6.0 behavior is not correct.
See this page for a simpler more academic version of the romberg method.
https://mathsci2.appstate.edu/~wmcb/Class/4310/ClassNotes131/Romberg_Program_Listings.pdf

I want to point out that this doesn't appear to be a possibly relevant numerical precision issue affecting the 7th or 9th decimal place.
The error is in the 1st or 2nd digit after the decimal (1 in 1000 or 1 in 10000), this is serious enough to cause the wrong conclusion to be reached.

I spent many hours studying the source code to _quadrature.py and I find no relevant changes that could cause an issue with the romberg method or any of the functions that it uses.

I also thought that this may be due to the depth of the recursion used in the romberg method, but that is also not the case.  It is more serious as the following screenshot shows:
![image](https://user-images.githubusercontent.com/80705677/111189605-df5fc800-857b-11eb-87c5-178337a13bc7.png)

#### Reproducing code example:

```
        cdfInvLog = np.array([-0.1, -0.15990408, -0.2734384,  -0.40681035, -0.55597077, -0.71510841,
        -5.63354942, -5.71273066, -5.71273066, -5.71273066, -5.80964067, -5.80964067,
        -5.80964067, -5.93457941, -5.93457941, -5.93457941, -6.41170067])
        a = -6.411700665769845
        b = -0.1
        from scipy.integrate import romberg
        from scipy.interpolate import interp1d

        ldpc_hard = 246
        f = interp1d(cdfInvLog, np.arange(0, len(cdfInvLog)))
        f2 = lambda y: ldpc_hard - f(y)

        fracArea = romberg(f2, a, b, show=True, rtol=5e-4, tol=5e-4, divmax=8)
        print("3377 - fracArea", type(fracArea), fracArea)
        expected = 1513.3537183      ### correct answer on scipy 1.5.4 and matches other implementations.
        self.assertEqual(expected, round(fracArea, 7))
        ### on scipy 1.6.0 we get the following answer: 1514.076528575215
        ### 1513.3537183 != 1514.0765286

```

#### Error message:

```

python -m unittest test_internal_taskflow.TestInternalTaskflow.test_run__romberg_integration_contrivedCase__test13820
34 - tadMysqlCfg/ debugVersion: |None| <class 'NoneType'>
Romberg integration of <function vectorize1.<locals>.vfunc at 0x7fac318491f0> from [-6.411700665769845, -0.1]

 Steps  StepSize   Results
     1  6.311701 1502.184759 
     2  3.155850 1510.022077 1512.634517 
     4  1.577925 1513.940737 1515.246956 1515.421119 
     8  0.788963 1515.900066 1516.553176 1516.640257 1516.659609 
    16  0.394481 1514.483561 1514.011393 1513.841941 1513.797523 1513.786299 
    32  0.197241 1514.171706 1514.067755 1514.071512 1514.075156 1514.076245 1514.076529 

The final result is 1514.076528575215 after 33 function evaluations.
3377 - fracArea <class 'numpy.float64'> 1514.076528575215
F
======================================================================
FAIL: test_run__romberg_integration_contrivedCase__test13820 (test_internal_taskflow.TestInternalTaskflow)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/home/src/etl/flows/tests/test_internal_taskflow.py", line 3426, in test_run__romberg_integration_contrivedCase__test13820
    self.assertEqual(expected, round(fracArea, 7))
AssertionError: 1513.3537183 != 1514.0765286

----------------------------------------------------------------------
Ran 1 test in 0.002s

FAILED (failures=1)

```

#### Scipy/Numpy/Python version information:

```
import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)
1.6.0 1.20.1 sys.version_info(major=3, minor=8, micro=6, releaselevel='final', serial=0)

```

**Comment by andyfaff** [comment]
Am investigating this now.

**Comment by andyfaff** [comment]
git history shows the `integrate.romberg` code hasn't changed in ages. git bisect points to 3c120b1e87ca52f246de5fb9d48fff9b0a1edd85. The problem with the test example provided is not an issue with `integrate.romberg`, rather it's an issue with the interpolation.

@mbrady-2, in your example `cdfInvLog` is reverse sorted. One of the steps that `interpolate.interp1d` takes is to sort into monotonically increasing values. Unfortunately `cdfInvLog` possesses several values that are repeated several times: `5.71273066`, `-5.80964067` and `-5.93457941`. Previously the `np.argsort` option was to do a `quicksort`, now it does a `stable` sort. A `stable` sort is where the order is not changed for values that are already sorted:

```
>>> np.argsort(cdfInvLog, kind='stable')    # new behaviour in 1.6.x
array([16, 13, 14, 15, 10, 11, 12,  7,  8,  9,  6,  5,  4,  3,  2,  1,  0])

>>> np.argsort(cdfInvLog)      # behaviour in 1.5.x, the quicksort is unstable
array([16, 14, 13, 15, 12, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1,  0])
```

These two argsorts are different, and will order `np.arange(0, len(cdfInvLog)` in different arrangements, thereby producing different `interpolate.interp1d` outputs. Your problem is ill-posed. For example, what should be the value of `f(-5.80964067)`, because there are three values for that entry?
This ill-posed problem cascades over into different values for the `romberg` integration. For this kind of problem I would recommend having unique values in `cdfInvLog`. Alternatively, you should ensure that the `x` values provided to `interp1d` are already sorted in monotonically increasing order (and use `assume_sorted=True`).

**Comment by mbrady-2** [comment]
@andyfaff You are correct.  Thanks so much for your prompt and thorough feedback.  
Using your suggestion, we were able to fix our code to correctly use scipy >= 1.6.0.

Thanks again,
Mike
